src/cpu/flexcpu/inflight_inst.hh

```diff
@@ -0,0 +1,592 @@
+    InstSeqNum _issueSeqNum;
+    StaticInstPtr instRef;
+
+    /**
+     * A storage for the exact PCState at the time of executing this
+     * instruction. Any changes to the PC through the instruction can also be
+     * captured in this variable, allowing this variable also to serve as the
+     * result storage for changes to the PCState. If this variable is changed
+     * in this fashion however, we may need to flush speculative state.
+     */
+    PCState _pcState;
+
+    Trace::InstRecord* _traceData = nullptr;
+
+    TimingRecord _timingRecord;
+
+    // Callbacks made during a state transition.
+    std::vector<std::function<void()>> commitCallbacks;
+    std::vector<std::function<void()>> completionCallbacks;
+
+    std::vector<std::function<void()>> effAddrCalculatedCallbacks;
+
+    // How many dependencies have to be resolved before I can execute?
+    size_t remainingDependencies = 0;
+
+    // How many dependencies have to be resolved before I can go to memory?
+    size_t remainingMemDependencies = 0;
+
+    std::vector<std::function<void()>> readyCallbacks;
+    std::vector<std::function<void()>> memReadyCallbacks;
+
+    std::vector<std::function<void()>> squashCallbacks;
+
+    std::vector<DataSource> sources;
+    std::vector<GenericReg> results;
+    std::vector<bool> resultValid;
+
+    /**
+     * For use in storing results of writes to miscRegs that don't go through
+     * the operands of the StaticInst.
+     */
+    std::vector<RegVal> miscResultVals;
+    std::vector<int> miscResultIdxs;
+
+    Fault _fault = NoFault;
+
+    // For use in tracking dependencies through memory
+    bool accessedPAddrsValid = false;
+    AddrRange accessedPAddrs;
+
+    bool _isSplitMemReq = false;
+    bool accessedSplitPAddrsValid = false;
+    AddrRange accessedSplitPAddrs; // Second variable to store range for second
+                                   // request as part of split accesses
+
+  public:
+    InflightInst(ThreadContext* backing_context, TheISA::ISA* backing_isa,
+                 MemIface* backing_mem_iface, InstSeqNum seq_num,
+                 const TheISA::PCState& pc_,
+                 StaticInstPtr inst_ref = StaticInst::nullStaticInstPtr);
+
+    // Unimplemented copy due to presence of raw pointers.
+    InflightInst(const InflightInst& other) = delete;
+
+    virtual ~InflightInst();
+
+    /**
+     * Tells this InflightInst to call a particular function when it has been
+     * made aware of a commit.
+     *
+     * NOTE: The order in which callbacks are added will determine the order in
+     *       which the callbacks are called upon notify. This is a hard promise
+     *       made by InflightInst, in order to provide some control on ordering
+     *       for any functionality where ordering is important.
+     *
+     * @param callback The function to call.
+     */
+    void addCommitCallback(std::function<void()> callback);
+    void addCommitDependency(InflightInst& parent);
```

Comment:
More comments would be useful here. You don't have to specify all of the params every time, but it would be good to say something like "This is the definition of dependency"

Also, it would be useful to have a comment saying when each dependency is resolved and when the callbacks are resolved. Is it after the stage (e.g., after commit for this one) or is it before, during? I can't ever remember...